doc/SAI-Proposal-VlanStack-QinQ.md

```diff
@@ -0,0 +1,380 @@
+Add an outer dot1q header with VID=100 for those packets with VID=10 at Ethernet10
+
+#### Usage
+
+```
+sai_object_id_t vlan_stacking_oid;
+sai_attribute_t attr;
+vector<sai_attribute_t> vlan_stacking_entry_attrs;
+
+attr.id = SAI_VLAN_STACK_ATTR_STAGE;
+attr.value.s32 = SAI_VLAN_STACK_STAGE_INGRESS;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_ACTION;
+attr.value.s32 = SAI_VLAN_STACK_ACTION_PUSH;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_MATCH_TYPE;
+attr.value.s32 = SAI_VLAN_STACK_MATCH_TYPE_INNER;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_ORIGINAL_VLAN_ID_LIST;
+sai_uint16_t vid_list[] = { 10 }
+attr.value.u16list.list = vid_list;
+attr.value.u16list.count = 1;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_PORT;
+attr.value.oid = 10;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_APPLIED_VLAN_ID;
+attr.value.u16 = 100;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_VLAN_APPLIED_PRI;
+attr.value.u8 = 0;
+vlan_stacking_entry_attrs.push(attr);
+
+sai_vlan_api->create_vlan_stack(&vlan_stacking_oid, gSwitchId, (uint32_t)vlan_stacking_entry_attrs.size(), vlan_stacking_entry_attrs.data());
+```
+
+#### Behavior
+
+![vlan_stack-ingress-push-inner](figures/vlan_stack/ingress_push_inner.png "Table 1: Ingress Push")
+__Table 1: Vlan Stack Ingress Push__
+
+### Egress/Remove outer tag based on outer tag
+Remove outer dot1q header for those packets with original outer dot1q header VID=100 at Ethernet20
+
+#### Usage
+
+```
+sai_object_id_t vlan_stacking_oid;
+sai_attribute_t attr;
+vector<sai_attribute_t> vlan_stacking_entry_attrs;
+
+attr.id = SAI_VLAN_STACK_ATTR_STAGE;
+attr.value.s32 = SAI_VLAN_STACK_STAGE_EGRESS;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_ACTION;
+attr.value.s32 = SAI_VLAN_STACK_ACTION_POP;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_MATCH_TYPE;
+attr.value.s32 = SAI_VLAN_STACK_MATCH_TYPE_OUTER;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_ORIGINAL_VLAN_ID_LIST;
+sai_uint16_t vid_list[] = { 100 }
+attr.value.u16list.list = vid_list;
+attr.value.u16list.count = 1;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_PORT;
+attr.value.oid = 20;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_VLAN_APPLIED_PRI;
+attr.value.u8 = 0;
+vlan_stacking_entry_attrs.push(attr);
+
+sai_vlan_api->create_vlan_stack(&vlan_stacking_oid, gSwitchId, (uint32_t)vlan_stacking_entry_attrs.size(), vlan_stacking_entry_attrs.data());
+```
+
+#### Behavior
+
+![vlan_stack-egress-pop-outer](figures/vlan_stack/egress_pop_outer.png "Table 2: Egress Pop Outer")
+__Table 2: Vlan Stack Egress Pop Outer__
+
+### Egress/Remove outer tag based on inner tag
+Remove outer dot1q header for those packets with original inner dot1q header VID=100 at Ethernet20
```

Comment:
Isn't this contradicting the SAI Egress pipeline? `Filter the packets based on first VID or PVID`
In this example we are matching "Inner" and popping "outer"

Reply:
This case might not be so useful. Since no people will want to send packets out of a port with different VID.
But in the behavior pipeline, this is a possible case. The packet will first be filtered by the port based on its VID, and do the vlan stacking afterwards.

The vlan stacking stage will be after the egress vlan filtering.

Reply:
I think I will remove this case in the next commit. Due to it is not so useful and also confuse people easily.